Add flag rendering to 'ClashOpts'

martijnbastiaan opened the pull request:
This adds a more structured way to define Clash command line flags to
`Clash.GHC.ClashFlags`. This comes with a couple of improvements over
the status quo:

  * Each boolean flag now defines two flags: `-fclash-foo` and
    `-fclash-no-foo`. This makes sure users can always unset a flag
    after setting it. This can be helpful in certain environments, and
    should help us once we'll support module-level flags.

  * Each flag backed by a `Maybe` can now be unset using `-fclash-no-foo`

Last but not least, `ClashOpts` can now be rendered to a set of flags.
This can help users of `clash-shake` to define a set of options using
`ClashOpts`, and then pass them to clash-the-binary.

## Still TODO:

  - [ ] Write a changelog entry (see changelog/README.md)
  - [ ] Check copyright notices are up to date in edited files
  - [ ] Document `*suffix` functions
  - [ ] Split up PR (=> lens stuff to separate one) to reduce noise?

martijnbastiaan committed:
Add lenses to `ClashOpts` and `DebugOpts`

alex-mckenna commented on clash-ghc/src-ghc/Clash/GHC/ClashFlag.hs:
```diff
@@ -0,0 +1,387 @@
+-- Types --
+-----------
+
+-- | Defines a way to parse command line flags, set their values in a 'ClashOpts',
+-- and render flags based on a 'ClashOpts'.
```
This is listed as three separate things, but I _think_ the parsing and setting are inherently related here. If that's the case it would be better to re-word and make it clear the two happen together / use bullet points

alex-mckenna commented on clash-ghc/src-ghc/Clash/GHC/ClashFlag.hs:
```diff
@@ -0,0 +1,387 @@
+-- | Defines a way to parse command line flags, set their values in a 'ClashOpts',
+-- and render flags based on a 'ClashOpts'.
+data ClashFlag = ClashFlag
+  { cfFlags :: IORef ClashOpts -> [Flag IO]
+  , cfRender :: ClashOpts -> [String]
```
Document these fields, it's kind of weird to think of a flag corresponding to multiple flags until you remember / know the functions later add `-fclash-foo` and `-fclash-no-foo` flags

alex-mckenna commented on clash-ghc/src-ghc/Clash/GHC/ClashFlag.hs:
```diff
@@ -0,0 +1,387 @@
+
+type FlagName = String
+
+--------------------------
+-- Simple flag builders --
+--------------------------
+
```
Marker comments like this in code are bad style IMO. They can end up in the wrong place, or have a label which is no longer relevant, and worse if a new function is added which could fit under more than one heading they make it harder to find things

alex-mckenna commented on clash-ghc/src-ghc/Clash/GHC/ClashFlag.hs:
```diff
@@ -0,0 +1,387 @@
+    addErr ("Could not parse: " <> s)
+    pure Nothing
+
+maybeIntFlag :: FlagName -> Lens' ClashOpts (Maybe Word) -> ClashFlag
+maybeIntFlag = maybeReadableFlag
+
+maybeWordFlag :: FlagName -> Lens' ClashOpts (Maybe Word) -> ClashFlag
+maybeWordFlag = maybeReadableFlag
```
I would just use `maybeReadableFlag` directly for these. Also `maybeIntFlag` has a `Maybe Word` which I assume should be `Maybe Int`

martijnbastiaan replied:
> Also maybeIntFlag has a Maybe Word which I assume should be Maybe Int

Woops

> I would just use maybeReadableFlag directly for these.

This is actually wrong in another way: ints are passed as `-fclash-foo=3`, but `maybeReadableFlag` uses `-fclash-foo 3`.

alex-mckenna replied:
> This is actually wrong in another way

Then I'm not sure I understand why they're defined like `maybeIntFlag = maybeReadableFlag` and `maybeWordFlag = maybeReadableFlag`. Either you can just use `maybeReadableFlag` or these definitions are wrong, no?

martijnbastiaan replied:
Yeah, the point of my comment was to point out my own code is wrong, not your comment. Sorry for the confusion.

> Either you can just use maybeReadableFlag or these definitions are wrong, no?

We've established a (frankly stupid) pattern that int/word is passed `-flike-this=3` while strings are passed `-flike-that str`. `maybeReadableFlag` interprets a string, so it will use the latter pattern. However, if we then use it to parse ints/words we're in the "wrong". So, it's not that `maybeReadableFlag` is wrong, but its usage for ints/words is.

I've put a comment somewhere that I think we should be liberal in what type of flag passing we accept for both.

alex-mckenna commented on clash-ghc/src-ghc/Clash/GHC/ClashFlag.hs:
```diff
@@ -0,0 +1,387 @@
+-- Generic flag builders --
+---------------------------
+
+-- | Helper function....
```
Comments like this aren't particularly helpful, _especially_ as haddock comments

martijnbastiaan replied:
Yeah it's a `TODO`

alex-mckenna commented on clash-ghc/src-ghc/Clash/GHC/ClashFlag.hs:
```diff
@@ -0,0 +1,387 @@
+    Nothing -> addErr ("Could not parse " <> s)  >> pure Nothing
+    a -> pure a
+
+---------------------------
+-- Generic flag builders --
+---------------------------
```
See previous comment on markers

alex-mckenna commented on clash-ghc/src-ghc/Clash/GHC/ClashFlags.hs:
```diff
@@ -1,327 +1,328 @@
+import           Data.List (intercalate)
+import           Text.Read (readMaybe)
+
+-- unordered-containers
+import qualified Data.Set as Set
+import           Data.Set (Set)
```
I'm also not a fan of these marker comments for package providence. Especially when it's wrong like this

alex-mckenna replied:
The main reason I'm against it is I don't see what benefit it brings. If you want to see if a package is unused there's `-Wunused-packages`, or `weeder` or any other tool which is going to be quicker and more correct than using these comments. If it's just because there's a lot of imports and it's difficult to read that's more a motivation to split up the module IMO

alex-mckenna commented on clash-ghc/src-ghc/Clash/GHC/ClashFlags.hs:
```diff
@@ -1,327 +1,328 @@
+  , deprecated "-fclash-float-support is always enabled from Clash 1.6 and onwards" $
+      boolFlag "float-support" voidLens
+
+  , deprecated "-fclash-clean has been removed" $
+      boolFlag "clean" voidLens
+
+  , deprecated "-fclash-nocache deprecated in favor of -fclash-no-cache" $
+      (boolFlag "nocache" (opt_cachehdl . iso' not not)){ cfRender = const [] }
```
I think much like deprecated APIs, when we deprecate a flag it should include a message saying it won't be a flag after version 1.X. Then when that version is released it gets removed and just becomes a normal "unknown flag" error

martijnbastiaan replied:
Agreed. Can this go in a separate PR? I don't want to battle out in which version things need to deprecate now :smile:

alex-mckenna replied:
The "default" is current + 2 major, I don't think that's controversial here?

martijnbastiaan replied:
Okay, I've set them all to 1.10 now. If this leads to any discussion I'm going to rip it from this PR.

alex-mckenna commented on clash-ghc/src-ghc/Clash/GHC/ClashFlags.hs:
```diff
@@ -1,327 +1,328 @@
+      -- TODO: Should we add a deprecation warning for these flags? We could
+      --       just use 'maybeReadableFlag' and ditch this custom impl if we
+      --       once the warning has been up long enough.
+      "Xilinx" -> pure (Just Vivado)
+      "ISE"    -> pure (Just Vivado)
+      "Altera" -> pure (Just Quartus)
+      "Intel"  -> pure (Just Quartus)
+      _ -> do
+        addErr (s ++ " is an unknown hdl synthesis tool")
+        pure Nothing
```
While we're here: I think we should

  * make this comparison case-insensitive, since it costs us nothing but potentially makes Clash less annoying
  * recognize the names `vivado` and `quartus` since it's weird to me we use those names internally but don't accept them on the command line

martijnbastiaan replied:
`Vivado` and `Quartus` are recognized. It only takes this branch if the `readMaybe` fails. Agreed on the lower case stuff, but this is true for all the "readable" flags. Let's postpone until we have a way to do those case-insensitively too?

alex-mckenna commented on clash-ghc/tests/Clash/Tests/GHC/ClashFlags.hs:
```diff
@@ -0,0 +1,407 @@
+  gen_dbg_transformationsFrom = genMaybe (opt_debug . dbg_transformationsFrom) (Gen.word Range.constantBounded)
+  gen_dbg_transformationsLimit = genMaybe (opt_debug . dbg_transformationsLimit) (Gen.word Range.constantBounded)
+
+  -- TODO
+  gen_dbg_transformations _fuzz = pure mempty
```
What is there TODO?

martijnbastiaan replied:
make it fuzz :-)

alex-mckenna commented on clash-ghc/tests/Clash/Tests/GHC/ClashFlags.hs:
```diff
@@ -0,0 +1,407 @@
+tests =
+  testGroup
+    "Clash.Tests.GHC.ClashFlags"
+    [ -- tasty-th doesn't account for Hedgehog tests
```
> `-- tasty-th doesn't account for Hedgehog tests`

Even if you prefix with `prop_` instead of `hprop_`, as they do in the hedgehog docs?

martijnbastiaan replied:
Yeah, it turns it into a `QuickCheck` test.

alex-mckenna commented on clash-lib/src/Clash/Driver/Types.hs:
```diff
@@ -138,59 +132,60 @@ data TransformationInfo
   --
   -- Command line flag: -fclash-debug-history[=FILE]
   } deriving (Generic, NFData, Show, Eq)
+makeLenses ''DebugOpts
```
I wonder if instead of changing the names to every field, it would be better to just call `makeLenses` with options, i.e. tell it to generate a lens `foo` from a field `dbg_foo`. If you do this here and for `ClashOpts` the diff will be much shorter

martijnbastiaan replied:
Don't we add `dbg_` to prevent naming collisions?

But even if we don't, `lens` is already complicated enough as it is, I think bringing in custom settings for smaller diffs isn't a good idea. The underscore-to-no-underscore pattern is pretty well established.

alex-mckenna replied:
> Don't we add dbg_ to prevent naming collisions?

No

> The underscore-to-no-underscore pattern is pretty well established.

Sure, it is more consistent to always use the same name scheme for lenses. It is a shame the default naming scheme is so intrusive though :cry:

martijnbastiaan replied:
It is. We'll have somewhat sensible records usable in 2030 though :-).

alex-mckenna commented on clash-lib/src/Clash/Driver/Types.hs:
```diff
@@ -540,6 +536,14 @@ newtype SdcInfo = SdcInfo
   { sdcClock :: [(Text, VDomainConfiguration)]
   }
 
+data ClashEnv = ClashEnv
+  { envOpts        :: ClashOpts
+  , envTyConMap    :: TyConMap
+  , envTupleTyCons :: IntMap TyConName
+  , envPrimitives  :: CompiledPrimMap
+  , envCustomReprs :: CustomReprs
+  } deriving (Generic, NFData)
+
```
Why did this move down here? You don't do anything with it after which would need the new lenses in scope

martijnbastiaan replied:
`TemplateHaskell` :sparkles:

martijnbastiaan commented on clash-ghc/src-ghc/Clash/GHC/ClashFlag.hs:
```diff
@@ -0,0 +1,387 @@
+{-|
+  Copyright   :  (C) 2021, QBayLogic B.V.
```
Nope, just living in the past :see_no_evil:

martijnbastiaan commented:
Thanks for the review @alex-mckenna, I'll implement the suggestions.

martijnbastiaan committed:
Refactor indention style of modules in `clash-ghc.cabal`

martijnbastiaan committed:
Add flag rendering to 'ClashOpts'

This adds a more structured way to define Clash command line flags to
`Clash.GHC.ClashFlags`. This comes with a couple of improvements over
the status quo:

  * Each boolean flag now defines two flags: `-fclash-foo` and
    `-fclash-no-foo`. This makes sure users can always unset a flag
    after setting it. This can be helpful in certain environments, and
    should help us once we'll support module-level flags.

  * Each flag backed by a `Maybe` can now be unset using `-fclash-no-foo`

Last but not least, `ClashOpts` can now be rendered to a set of flags.
This can help users of `clash-shake` to define a set of options using
`ClashOpts`, and then pass them to clash-the-binary.
